dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,597 @@
+
+        super(HyperbandSearchCV, self).__init__(
+            estimator,
+            param_distribution,
+            max_iter=max_iter,
+            patience=patience,
+            tol=tol,
+            test_size=test_size,
+            random_state=random_state,
+            scoring=scoring,
+        )
+
+    def _get_SHAs(self, brackets):
+
+        patience = _get_patience(self.patience, self.max_iter, self.aggressiveness)
+
+        # This is the first time self.random_state is used after
+        # HyperbandSearchCV.fit is called.
+        seed_start = check_random_state(self.random_state).randint(2 ** 31)
+        self._SHA_seed = seed_start
+
+        # These brackets are ordered by adaptivity; bracket=0 is least adaptive
+        SHAs = [
+            (
+                b,
+                SuccessiveHalvingSearchCV(
+                    self.estimator,
+                    self.param_distribution,
+                    n_initial_parameters=n,
+                    start_iter=r,
+                    adaptive_max_iter=b + 1,
+                    aggressiveness=self.aggressiveness,
+                    max_iter=self.max_iter,
+                    patience=patience,
+                    tol=self.tol,
+                    test_size=self.test_size,
+                    random_state=seed_start + b if b != 0 else self.random_state,
+                    scoring=self.scoring,
+                ),
+            )
+            for b, (n, r) in brackets.items()
+        ]
+        return SHAs
+
+    @gen.coroutine
+    def _fit(self, X, y, **fit_params):
+        X = self._check_array(X)
+        y = self._check_array(y, ensure_2d=False)
+        scorer = check_scoring(self.estimator, scoring=self.scoring)
+
+        brackets = _get_hyperband_params(self.max_iter, eta=self.aggressiveness)
+
+        SHAs = self._get_SHAs(brackets)
+        # Which bracket to run first? Going to go with most adaptive;
+        # hopefully less adaptive can fill in for any blank spots
+        #
+        # _brackets is ordered from largest to smallest
+        _brackets_ids = [b for b, SHA in SHAs]
+        _SHAs = yield [SHA.fit(X, y, **fit_params) for b, SHA in SHAs]
+        SHAs = {b: SHA for b, SHA in zip(_brackets_ids, _SHAs)}
+
+        # This for-loop rename estimator IDs and pulls out wall times
+        key = lambda b, old: "bracket={}-{}".format(b, old)
+        for b, SHA in SHAs.items():
+            new_ids = {old: key(b, old) for old in SHA.cv_results_["model_id"]}
+            SHA.cv_results_["model_id"] = np.array(
+                [new_ids[old] for old in SHA.cv_results_["model_id"]]
+            )
+            SHA.model_history_ = {
+                new_ids[old]: v for old, v in SHA.model_history_.items()
+            }
+            for hist in SHA.model_history_.values():
+                for h in hist:
+                    h["model_id"] = new_ids[h["model_id"]]
+                    h["bracket"] = b
+
+        keys = list(SHA.cv_results_.keys())
```

Comment:
Is `SHA` correct here? That's a name that's leaking from the previous for loop (so a potential NameError if `SHAs` is empty.

Reply:
I've changed this to `SHAs[0]`, and have implemented a test to make sure SHAs is always non-empty.